File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
       
        if not list1 and not list2 :
            return None
        
        if not list1:
            return list2
        
        if not list2:
            return list1
        
        newList = ListNode()
        
        if list1.val < list2.val :
            newList = list1
            list1 = list1.next
            
        else:
            newList = list2
            list2 = list2.next
            
        ans = newList
        
        while list1 and list2:
            

            if  list1.val < list2.val:
                newList.next =   list1
                list1 = list1.next
            
            else:
                newList.next =   list2
                list2 = list2.next
                
            newList = newList.next
        if list1:
            newList.next = list1
        else:
            newList.next =list2
            
        
        return ans
            

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(!l1 && !l2){
            
            return nullptr;
        }
        if(!l1)
            return l2;
        if(!l2)
            return l1;
        
        ListNode * head = nullptr;
        
        if(l1->val <= l2->val){
            head = l1;
            l1 = l1->next;
        }
        else{
            head = l2;
            l2 = l2->next;
        }
        
        ListNode *iter = head;
        
        
        while(l1 && l2){
         
            if(l1->val <= l2->val){
                iter->next = l1;
                l1 = l1->next;
            }
            else{
                iter->next = l2;
                l2 = l2->next;
            } 
            iter = iter->next;
        }
        
        if(l1)
            iter->next = l1;
        else
            iter->next = l2;
        
        
        
        return head;
    }
};